### PR TITLE
Add changes indexer

### PR DESCRIPTION
**Pull request by sdlyy**
Resolves L2B-2877

## What's changed
- Added indexer for changelog
- Pairs of outputs are compared and transformed into granular entries
- Changes indexer relies on discovery indexer
- Contracts can be added/removed - those are represented as milestones to track contract deployments
- Contract's properties are tracked via changelog entries in a separate table
- Each chain configuration now have changelog whitelist - only listed addresses will be taken into account while performing comparison
- There are quite a few tests :trollface: and comments, feel free to question

## Notes
- Logic might be simplified and not split into buckets (`added`, `removed`, `modified`). Also, those buckets are defined from contract standpoint - not the property. Within a `modified` bucket, you won't only see properties that have been modified. It simply means that the given contract is not a newly added one. 

## How to test it?
- Prune database
- Pull latest production dump
- Enable any chain you would like to perform diffing for
- Run application on fresh data to perform migration
- Once indexer runs, take a peak into milestones and changelog tables to see the actual changes that have been gathered based on already present discovery outputs

## Result
image
image

## Done
- [ ] Indexer test
- [x] Repositories' tests
- [x] Test for milestones
- [x] Cleanup getChangelogChanges - maybe split, maybe not
- [x] Cleanup indexer
- [x] Add test when contract has not values - wut?
- [x] Test outstanding funcs
- [x] Ignore list of changes - optional matches

### CODE REVIEW

File: packages/backend/src/config/discovery/arbitrum.ts [hunk 1]
```diff
@@ -19,3 +27,7 @@ const arbitrumRawConfig = createConfigFromTemplate({
 
 const arbitrumDiscoveryConfig = new DiscoveryConfig(arbitrumRawConfig)
 const arbitrumEventsToWatch = getEventsToWatch(addresses)
+const arbitrumChangelogWhitelist = toEthereumAddresses([
+  addresses.ultraLightNodeV2,
+  addresses.endpoint,
+])
```
Comment (michalsidzej) [inline packages/backend/src/config/discovery/arbitrum.ts:33]: I think we also want to include changes in the msig

File: packages/backend/src/indexers/ChangelogIndexer.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    discoveryIndexer: DiscoveryIndexer,
+    logger: Logger,
+  ) {
+    super(logger.tag(ChainId.getName(chainId)), [discoveryIndexer])
+  }
+
+  override async update(from: number, to: number): Promise<number> {
+    const fromBlockRecord = await this.blockNumberRepository.findAtOrBefore(
+      new UnixTime(from),
+      this.chainId,
```
Comment (michalsidzej) [inline packages/backend/src/indexers/ChangelogIndexer.ts:40]: Please update the indexer with new Indexer logic (block number as height)

**Review by michalsidzej — CHANGES_REQUESTED** [review]
Looks logical to me. Please update the Indexer logic

File: packages/backend/src/config/discovery/arbitrum.ts (hunk 1 above)
Comment (sdlyy) [reply packages/backend/src/config/discovery/arbitrum.ts]: https://github.com/l2beat/lz-monitoring/pull/56/commits/45b50157501d0f47936e0001a2cf853a1f3fb273

File: packages/backend/src/indexers/ChangelogIndexer.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private readonly discoveryRepository: DiscoveryRepository,
+    private readonly chainId: ChainId,
+    private readonly changelogWhitelist: EthereumAddress[],
+    discoveryIndexer: DiscoveryIndexer,
+    logger: Logger,
+  ) {
+    super(logger.tag(ChainId.getName(chainId)), [discoveryIndexer])
+  }
+
+  override async update(from: number, to: number): Promise<number> {
```
Comment (michalsidzej) [inline packages/backend/src/indexers/ChangelogIndexer.ts:35]:
```suggestion
  override async update(fromBlockNumber: number, toBlockNumber: number): Promise<number> {
```

File: packages/backend/src/indexers/ChangelogIndexer.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    const discovery = await this.discoveryRepository.getSortedInRange(
+      fromBlockNumber,
+      toBlockNumber,
+      this.chainId,
+    )
+
+    // 0 records - we can't compare to anything
+    if (discovery.length === 0) {
+      return to
```
Comment (michalsidzej) [inline packages/backend/src/indexers/ChangelogIndexer.ts:48]:
```suggestion
      return toBlockNumber
```

File: packages/backend/src/indexers/ChangelogIndexer.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    const changes = outputPairs.map(([previousOutput, currentOutput]) =>
+      getDiscoveryChanges(previousOutput, currentOutput),
+    )
+
+    const flatChanges = flattenChanges(changes)
+
+    await this.changelogRepository.addMany(flatChanges.properties)
+    await this.milestoneRepository.addMany(flatChanges.milestones)
+
+    return to
```
Comment (michalsidzej) [inline packages/backend/src/indexers/ChangelogIndexer.ts:87]:
```suggestion
    return toBlockNumber
```

Comment (sdlyy) [reply packages/backend/src/indexers/ChangelogIndexer.ts]: https://github.com/l2beat/lz-monitoring/pull/56/commits/d939e61e33984dcdb63ea9876c748e7899fe5e7f

File: packages/backend/src/indexers/ChangelogIndexer.ts (hunk 4 above)
Comment (sdlyy) [reply packages/backend/src/indexers/ChangelogIndexer.ts]: https://github.com/l2beat/lz-monitoring/pull/56/commits/d939e61e33984dcdb63ea9876c748e7899fe5e7f

File: packages/backend/src/indexers/ChangelogIndexer.ts (hunk 3 above)
Comment (sdlyy) [reply packages/backend/src/indexers/ChangelogIndexer.ts]: https://github.com/l2beat/lz-monitoring/pull/56/commits/d939e61e33984dcdb63ea9876c748e7899fe5e7f

**Review by michalsidzej — APPROVED** [review]

**Review by michalsidzej — APPROVED** [review]

**Pull request merged**